src/volume_calc.cpp

```diff
@@ -90,182 +123,236 @@ std::vector<VolumeCalculation::Result> VolumeCalculation::execute() const
-    if (mpi::master) {
-      for (int j = 1; j < mpi::n_procs; j++) {
-        int q;
-        MPI_Recv(&q, 1, MPI_INTEGER, j, 0, mpi::intracomm, MPI_STATUS_IGNORE);
+      if (mpi::master) {
+        for (int j = 1; j < mpi::n_procs; j++) {
+          int q;
+          MPI_Recv(&q, 1, MPI_INTEGER, j, 0, mpi::intracomm, MPI_STATUS_IGNORE);
+          int buffer[2*q];
+          MPI_Recv(&buffer[0], 2*q, MPI_INTEGER, j, 1, mpi::intracomm, MPI_STATUS_IGNORE);
+          for (int k = 0; k < q; ++k) {
+            for (int m = 0; m < master_indices[i_domain].size(); ++m) {
+              if (buffer[2*k] == master_indices[i_domain][m]) {
+                master_hits[i_domain][m] += buffer[2*k + 1];
+                break;
+              }
+            }
+          }
+        }
+      } else {
+        int q = master_indices[i_domain].size();
         int buffer[2*q];
-        MPI_Recv(&buffer[0], 2*q, MPI_INTEGER, j, 1, mpi::intracomm, MPI_STATUS_IGNORE);
         for (int k = 0; k < q; ++k) {
-          for (int m = 0; m < master_indices[i_domain].size(); ++m) {
-            if (buffer[2*k] == master_indices[i_domain][m]) {
-              master_hits[i_domain][m] += buffer[2*k + 1];
+          buffer[2*k] = master_indices[i_domain][k];
+          buffer[2*k + 1] = master_hits[i_domain][k];
+        }
+
+        MPI_Send(&q, 1, MPI_INTEGER, 0, 0, mpi::intracomm);
+        MPI_Send(&buffer[0], 2*q, MPI_INTEGER, 0, 1, mpi::intracomm);
+      }
+#endif
+
+      if (mpi::master) {
+        int total_hits = 0;
+        for (int j = 0; j < master_indices[i_domain].size(); ++j) {
+          total_hits += master_hits[i_domain][j];
+          double f = static_cast<double>(master_hits[i_domain][j]) / total_samples;
+          double var_f = f*(1.0 - f) / total_samples;
+
+          int i_material = master_indices[i_domain][j];
+          if (i_material == MATERIAL_VOID) continue;
+
+          const auto& mat = model::materials[i_material];
+          for (int k = 0; k < mat->nuclide_.size(); ++k) {
+            // Accumulate nuclide density
+            int i_nuclide = mat->nuclide_[k];
+            atoms(i_nuclide, 0) += mat->atom_density_[k] * f;
+            atoms(i_nuclide, 1) += std::pow(mat->atom_density_[k], 2) * var_f;
+          }
+        }
+
+        // Determine volume
+        result.volume[0] = static_cast<double>(total_hits) / total_samples * volume_sample;
+        result.volume[1] = std::sqrt(result.volume[0]
+          * (volume_sample - result.volume[0]) / total_samples);
+        result.iterations = iterations;
+
+        // update threshold value if needed
+        if (trigger_type_ != VolumeTriggerMetric::NONE) {
+          double val = 0.0;
+          switch (trigger_type_) {
+            case VolumeTriggerMetric::STD_DEV:
+              val = result.volume[1];
+              break;
+            case VolumeTriggerMetric::REL_ERR:
+              val = result.volume[0] == 0.0 ? 0.0 : result.volume[1] / result.volume[0];
```

Comment:
So hypothetically if a region gets no hits at all (maybe too few samples or something), it seems like this could result in the calculation finishing since the relative error will be below the threshold. Should the true case be infinity instead of 0.0?

Reply:
Great catch, thanks! That was leftover from when I wasn't properly resetting `trigger_val`.